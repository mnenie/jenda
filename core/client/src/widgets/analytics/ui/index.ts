import BoardsChart from "./BoardsChart.vue";
import OnlineChart from "./OnlineChart.vue";
import SharedSection from "./SharedSection.vue";
import TasksChart from "./TasksChart.vue";
import UsersChart from "./UsersChart.vue";

export { BoardsChart, OnlineChart, SharedSection, TasksChart, UsersChart };
