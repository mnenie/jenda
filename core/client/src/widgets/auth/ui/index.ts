import AuthContainer from './AuthContainer.vue'

export { AuthContainer }
