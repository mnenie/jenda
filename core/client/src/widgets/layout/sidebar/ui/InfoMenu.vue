<script setup lang="ts">
import { ref } from 'vue'
import { createReusableTemplate } from '@vueuse/core'
import { useExpanded } from '@/shared/composables'
import {
  UiBadge,
  UiButton,
  UiDropdownMenu,
  UiDropdownMenuContent,
  UiDropdownMenuGroup,
  UiDropdownMenuItem,
  UiDropdownMenuTrigger,
} from '@/shared/ui'
import { HotkeysDialog } from '@/widgets/dialogs'

defineProps<{
  isShowPlan?: boolean
}>()

const expanded = useExpanded()
const { isExpanded } = expanded.getExpanded()

const modalOpen = ref(false)

const [DifineTemplate, ReuseTemplate] = createReusableTemplate()
</script>

<template>
  <DifineTemplate v-slot="{ title, icon }">
    <UiDropdownMenuItem>
      <span class="text-sm" :class="icon" />
      <span class="2xl:text-13px text-sm font-medium">{{ title }}</span>
    </UiDropdownMenuItem>
  </DifineTemplate>
  <UiDropdownMenu direction="right">
    <UiDropdownMenuTrigger as-child>
      <UiButton
        v-tooltip.right="{
          content: $t(`sidebar.help.title'`),
          triggers: ['hover'],
          disabled: isExpanded,
        }"
        variant="ghost"
        class="w-full gap-2 shadow-none py-0 px-2 transition-all duration-200 ease justify-between mb-1"
        :class="[isExpanded ? 'justify-between' : 'justify-center p-0']"
      >
        <div class="flex items-center gap-2">
          <div i-hugeicons-mortarboard-02 class="!w-15px !h-15px 2xl:(!w-4 !h-4) text-neutral-800 dark:text-neutral-300" />
          <span v-show="isExpanded" class="text-13px 2xl:text-sm !fw500 text-neutral-900 dark:text-neutral-100">
            {{ $t('sidebar.help.title') }}
          </span>
        </div>
        <UiBadge v-if="isShowPlan" variant="soft" class="w-10 px-1">
          <div class="flex items-center gap-0.5">
            <div i-lucide-stars />
            <span class="!text-10px !fw500">1/3</span>
          </div>
        </UiBadge>
      </UiButton>
    </UiDropdownMenuTrigger>
    <UiDropdownMenuContent align="center" side="right" class="w-102%">
      <UiDropdownMenuGroup>
        <ReuseTemplate
          :title="$t('sidebar.help.items', 0)"
          icon="i-hugeicons-agreement-01"
        />
        <ReuseTemplate
          :title="$t('sidebar.help.items', 1)"
          icon="i-hugeicons-book-open-01"
          @click="modalOpen = true"
        />
        <ReuseTemplate
          :title="$t('sidebar.help.items', 2)"
          icon="i-hugeicons-link-forward"
        />
      </UiDropdownMenuGroup>
    </UiDropdownMenuContent>
  </UiDropdownMenu>
  <HotkeysDialog v-model:open="modalOpen" />
</template>
