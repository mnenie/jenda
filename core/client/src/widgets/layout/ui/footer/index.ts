import AppFooter from './AppFooter.vue'

export { AppFooter }
