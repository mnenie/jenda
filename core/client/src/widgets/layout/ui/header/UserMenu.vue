<script setup lang="ts">
import { RouteNames } from '@/shared/config/consts'
import {
  UiDropdownMenu,
  UiDropdownMenuContent,
  UiDropdownMenuGroup,
  UiDropdownMenuItem,
  UiDropdownMenuLabel,
  UiDropdownMenuSeparator,
  UiDropdownMenuShortcut,
  UiDropdownMenuTrigger,
} from '@/shared/ui'
</script>

<template>
  <UiDropdownMenu direction="right">
    <UiDropdownMenuTrigger as-child>
      <div class="flex items-center justify-center h-17px">
        <div
          i-lucide-ellipsis-vertical
          class="text-base text-neutral-500 dark:text-neutral-400"
        />
      </div>
    </UiDropdownMenuTrigger>
    <UiDropdownMenuContent align="end" class="w-[200px] mt-4px">
      <UiDropdownMenuLabel class="2xl:text-[13px] sm:text-sm">
        test@gmail.com
      </UiDropdownMenuLabel>
      <UiDropdownMenuSeparator />
      <UiDropdownMenuGroup>
        <UiDropdownMenuItem @click="$router.push({ name: RouteNames.welcome })">
          <span class="2xl:text-xs text-sm font-medium">{{ $t('header.user.welcome') }}</span>
          <UiDropdownMenuShortcut>⇧P</UiDropdownMenuShortcut>
        </UiDropdownMenuItem>
        <UiDropdownMenuItem>
          <span class="2xl:text-xs text-sm font-medium">{{ $t('header.user.logout') }}</span>
          <UiDropdownMenuShortcut>⌘X</UiDropdownMenuShortcut>
        </UiDropdownMenuItem>
      </UiDropdownMenuGroup>
    </UiDropdownMenuContent>
  </UiDropdownMenu>
</template>
