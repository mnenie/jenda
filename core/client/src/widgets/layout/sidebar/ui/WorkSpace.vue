<script setup lang="ts">
import { computed } from 'vue'
import { Icon } from '@iconify/vue'
import { useRoute } from 'vue-router/auto'
import type { WorkspaceLink } from '@/shared/config/types-shared'
import { useExpanded } from '@/shared/composables'
import { UiBadge, UiButton } from '@/shared/ui'

const props = defineProps<{
  links: WorkspaceLink[]
}>()

const route = useRoute()

function isCurrentPath(link: WorkspaceLink): boolean {
  if (route.name === 'boards') {
    return link.pathName === 'boards'
  }
  if (route.name === 'community') {
    return link.pathName === 'templates'
  }
  return route.path === `/${link.pathName}`
}

const pathName = computed(() => {
  return props.links.map(link => ({
    ...link,
    isActive: isCurrentPath(link),
  }))
})

const expanded = useExpanded()

const { isExpanded } = expanded.getExpanded()

const contentPosition = computed(() => {
  return isExpanded.value ? 'space-between' : 'center'
})
</script>

<template>
  <div class="flex flex-col space-y-1.5 justify-start mb-20px">
    <RouterLink
      v-for="link in pathName"
      :key="link.id"
      v-tooltip.right="{
        content: $t(`sidebar.${link.name}`),
        triggers: ['hover'],
        disabled: isExpanded,
      }"
      :to="`/${link.pathName}`"
      class="flex items-center gap-1.5 cursor-pointer text-neutral-900 justify-start w-full"
    >
      <UiButton
        :variant="link.isActive ? 'secondary' : 'ghost'"
        size="sm"
        class="w-full gap-2 shadow-none px-2 transition-all duration-200 ease"
        :class="link.isActive ? '!bg-neutral-200/40 dark:!bg-neutral-700/50' : ''"
        :style="{ padding: !isExpanded ? '0px' : '', justifyContent: contentPosition }"
      >
        <div class="flex items-center gap-2">
          <Icon :icon="link.icon" class="!w-15px !h-15px 2xl:(!w-4 !h-4) text-neutral-900 dark:text-neutral-200" />
          <span v-show="isExpanded" class="text-13px 2xl:text-sm !fw500 text-neutral-900 dark:text-neutral-100">{{ $t(`sidebar.${link.name}`) }}</span>
        </div>
        <UiBadge v-if="link.id === 1 && isExpanded" variant="solid" class="px-5px py-0 text-xs rounded-md">
          <span class="text-[10px] text-white">3</span>
        </UiBadge>
      </UiButton>
    </RouterLink>
  </div>
</template>
