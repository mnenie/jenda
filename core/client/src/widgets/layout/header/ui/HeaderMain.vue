<script setup lang="ts" generic="T extends Board">
import { computed, shallowReactive, toRef } from 'vue'
import { Icon } from '@iconify/vue'
import UserMenu from './UserMenu.vue'
import type { Board } from '@/entities/board'
import type { ProjectLink } from '@/shared/config/types-shared'
import { type User, UserAvatar } from '@/entities/user'
import { useLayoutPaths } from '@/shared/composables'
import { links } from '@/shared/constants'
import { HotkeysDialog, ShareDialog } from '@/widgets/dialogs'
import { ShimmerButton } from '@/shared/ui'

const props = defineProps<{
  projects: Board[]
}>()

const users = shallowReactive<User[]>([
  {
    _id: '0',
    email: 'alex@gmail.com',
    photoUrl: 'https://avatars.githubusercontent.com/u/121057011?v=4',
  },
  {
    _id: '1',
    email: 'airat@gmail.com',
    photoUrl: 'https://avatars.githubusercontent.com/u/95149637?s=100&v=4',
  },
  {
    _id: '2',
    email: 'slava@gmail.com',
    photoUrl: 'https://avatars.githubusercontent.com/u/83920644?v=4',
  },
  {
    _id: '3',
    email: 'egor@gmail.com',
    photoUrl: 'https://avatars.githubusercontent.com/u/84976745?v=4',
  },
  {
    _id: '4',
    email: 'egor@gmail.com',
    photoUrl: 'https://avatars.githubusercontent.com/u/84976745?v=4',
  },
  {
    _id: '5',
    email: 'egor@gmail.com',
    photoUrl: 'https://avatars.githubusercontent.com/u/84976745?v=4',
  },
])

const visibleUsers = computed(() => users.slice(0, 3))

const remaining = computed(() => Math.max(users.length - 3, 0))

const userPosition = computed(() => (index: number) => (visibleUsers.value.length - index) * 20)

const _projects = toRef(props, 'projects') as unknown as ProjectLink[]
const { active } = useLayoutPaths(links, _projects)
</script>

<template>
  <header
    class="sticky top-0 left-0 right-0 z-999 w-full flex items-center justify-between px-15px py-8px min-h-52px max-h-52px
    border-b border-b-solid border-layout"
  >
    <div class="inline-flex items-center gap-3">
      <Icon
        v-if="active.extendedAttrs.icon"
        :icon="active.extendedAttrs.icon"
        class="text-20px text-neutral-800 dark:text-neutral-200"
      />
      <div
        v-if="active.extendedAttrs.color"
        i-jenda-project
        :style="{ color: active.extendedAttrs.color, fontSize: '20px' }"
      />
      <p v-if="active" class="text-lg mt-1.5px !fw-600 dark:text-neutral-100">
        {{ active.name }}
      </p>
    </div>
    <div />
    <div class="flex items-center h-full gap-2.5">
      <div class="relative flex items-center">
        <UserAvatar
          v-if="remaining > 0"
          :style="{
            right: `0px`,
            zIndex: `${visibleUsers.length + 1}`,
          }"
          class="!h-28px !w-28px absolute flex items-center justify-center text-sm"
        >
          +{{ remaining }}
        </UserAvatar>
        <template v-for="user, idx in visibleUsers" :key="user._id">
          <UserAvatar
            :style="{
              right: `${userPosition(idx)}px`,
              zIndex: `${idx + 1}`,
            }"
            class="!h-28px !w-28px absolute"
          >
            <img :src="user.photoUrl" width="100%" />
          </UserAvatar>
        </template>
      </div>
      <ShareDialog>
        <ShimmerButton shimmer-size="2px">
          <div i-hugeicons-link-04 class="text-neutral-900 dark:text-neutral-100 w-16px h-16px 2xl:(w-4 h-4)" />
          <span
            class="whitespace-pre-wrap text-center text-neutral-800 dark:text-neutral-100 text-sm fw500"
          >
            {{ $t('header.share') }}
          </span>
        </ShimmerButton>
      </ShareDialog>
      <div
        v-tooltip="{ content: $t('header.navigator.messages'), trigger: ['hover'], distance: 7 }"
        i-hugeicons-message-multiple-01
        class="text-base text-neutral-700 cursor-pointer outline-none dark:text-neutral-300"
        style="margin-right: 2px"
      />
      <HotkeysDialog>
        <div
          v-tooltip="{ content: $t('header.navigator.question'), trigger: ['hover'], distance: 7 }"
          i-hugeicons-book-open-01
          class="text-base text-neutral-700 cursor-pointer outline-none dark:text-neutral-300"
          style="margin-right: -1px;"
        />
      </HotkeysDialog>
      <UserMenu />
    </div>
  </header>
</template>
