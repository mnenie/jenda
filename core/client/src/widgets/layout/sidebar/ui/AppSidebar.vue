<script setup lang="ts">
import { computed } from 'vue'
import { refDebounced } from '@vueuse/core'
import InfoMenu from './InfoMenu.vue'
import IntegrationItems from './IntegrationItems.vue'
import ProjectsList from './ProjectsList.vue'
import SearchBox from './SearchBox.vue'
import WorkSpace from './WorkSpace.vue'
import WorkspaceChooser from './WorkSpaceChooser.vue'
import type { Board } from '@/entities/board'
import { PlanCard } from '@/widgets/plan'
import { links } from '@/shared/constants'
import { useExpanded } from '@/shared/composables'

defineProps<{
  boards: Board[]
}>()

const expanded = useExpanded()

const { isExpanded } = expanded.getExpanded()

const debounced = refDebounced(isExpanded, 20)

const isShowPlan = computed((): boolean => {
  return isExpanded.value ? debounced.value : false
})

const paddingExpanded = computed(() => {
  return isExpanded.value ? '10px' : '0px 10px 10px 10px'
})
</script>

<template>
  <div class="relative w-full flex flex-col h-full bg-sidebar select-none border-r border-r-solid border-layout">
    <WorkspaceChooser />
    <div
      class="flex flex-col justify-between h-full mt-1.5"
      :style="{ padding: paddingExpanded }"
    >
      <div>
        <SearchBox :boards />
        <WorkSpace :links />
        <IntegrationItems />
        <ProjectsList :boards />
      </div>
      <div>
        <InfoMenu :is-show-plan />
        <PlanCard :is-show-plan />
      </div>
    </div>
  </div>
</template>
