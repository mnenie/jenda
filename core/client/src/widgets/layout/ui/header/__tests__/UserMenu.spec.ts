import i18n from "@/shared/lib/i18n";
import { UiDropdown, UiDropdownItem } from "@/shared/ui";
import { mount } from "@vue/test-utils";
import { describe, expect, it, vi } from "vitest";
import { type ComponentPublicInstance, nextTick } from "vue";
import UserMenu from "../UserMenu.vue";
import "@/shared/lib/vitest-utils/cookiesI18n-mock";

type UserMenuDropdownInstance = ComponentPublicInstance<
  {},
  {},
  {
    handleDropdown: () => void;
  }
>;

const mockRouter = {
  push: vi.fn(),
  beforeEach: vi.fn(),
};

describe("tests for UserMenu.vue", () => {
  const wrapper = mount(UserMenu, {
    global: {
      plugins: [i18n],

      mocks: {
        t: (key: string) => {
          const translations: Record<string, string> = {
            "header.user.welcome": "welcome",
            "header.user.logout": "logout",
          };
          return translations[key];
        },
        $router: mockRouter,
      },
    },
  });

  it("should be render correctly", async () => {
    wrapper
      .findComponent(UiDropdown as unknown as UserMenuDropdownInstance)
      .vm.handleDropdown();
    await nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should redirect coorectly", async () => {
    const welcomeTrigger = wrapper
      .find(".content")
      .findAllComponents(UiDropdownItem)
      .at(0);

    await welcomeTrigger.trigger("click");
    expect(mockRouter.push).toHaveBeenCalledWith({ name: "welcome" });
  });
});
