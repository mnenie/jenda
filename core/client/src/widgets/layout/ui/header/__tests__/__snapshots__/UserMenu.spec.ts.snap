// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for UserMenu.vue > should be render correctly 1`] = `
"<div class="dropdown">
  <div class="wrapper">
    <div class="wrapper">
      <div i-lucide-ellipsis-vertical="" class="icon" text-base=""></div>
    </div>
  </div>
  <transition-stub name="dropdown" appear="false" persisted="false" css="true">
    <div class="inside" style="top: 36px; right: 0px; transform: translateX(0%);">
      <div class="text-sm header"><span class="text-sm" style="font-weight: 500;">test@gmail.com</span></div>
      <div class="content">
        <div class="item">Go to Welcome <div class="shortcut">⌘B</div>
        </div>
        <div class="item">Log out <div class="shortcut">⇧⌘Q</div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>"
`;
